_posts/2023-04-19-more-optimizations.md

```diff
@@ -0,0 +1,1300 @@
+architecture, 14 native instructions are needed.
+
+If the type of the operands are known to be integers sufficiently
+small making overflow impossible, the JIT needs to emit only 5 native
+instructions for the addition.
+
+Here is an example where the types and ranges of the operands for the
+`+` operator are known:
+
+```erlang
+add5(X, Y) when X =:= X band 16#3FF,
+                Y =:= Y band 16#3FF ->
+    X + Y.
+```
+
+This code results in the following BEAM code for the `+` operator:
+
+```
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,{0,1023}}},{tr,{x,1},{t_integer,{0,1023}}}],
+            {x,0}}.
+```
+
+The register operands (`{x,0}` and `{x,1}`) have now been annotated with
+type information:
+
+```erlang
+{tr,Register,Type}
+```
+
+That is, each register operand is a three-tuple with `tr` as the first
+element. `tr` stands for **typed register**. The second element is the
+BEAM register (`{x,0}` or `{x,1}` in this case), and the third element
+is the type of the register in the compiler's internal type
+representation. `{t_integer,{0,1023}}` means that the value is an
+integer in the inclusive range 0 through 1023.
+
+With that type information, the JIT emits the following native code
+for the `+` operator:
+
+```nasm
+# i_plus_ssjd
+# add without overflow check
+    mov rax, qword ptr [rbx]
+    mov rsi, qword ptr [rbx+8]
+    and rax, -16               ; Zero the tag bits
+    add rax, rsi
+    mov qword ptr [rbx], rax
+```
+
+(Lines starting with `#` are comments emitted by the JIT, while the
+text that follows `;` is a comment added by me for clarification.)
+
+The reduction in code size from 14 instructions down to 5 is nice, but
+having to express the range check in that convoluted way using `band`
+can hardly be called nice nor natural.
+
+If we try to express the range checks in a more natural way:
+
+```erlang
+add4(X, Y) when is_integer(X), 0 =< X, X < 16#400,
+                is_integer(Y), 0 =< Y, Y < 16#400 ->
+    X + Y.
+```
+
+the compiler in OTP 25 will no longer be able to figure out the
+ranges for the operands:
+
+```
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,any}},{tr,{x,1},{t_integer,any}}],
+            {x,0}}.
+```
+
+Because of that severe limitation in the compiler's value range
+analysis, I wrote:
+
+> We aim to improve the type analysis and optimizations in OTP 26 and
+generate better code for this example.
+
+### The enhanced type-based optimizations in OTP 26
+
+Compiling the same example with OTP 26, the result is:
+
+
+```
+    {test,is_integer,{f,19},[{x,0}]}.
+    {test,is_ge,{f,19},[{tr,{x,0},{t_integer,any}},{integer,0}]}.
+    {test,is_ge,{f,19},[{integer,1023},{tr,{x,0},{t_integer,{0,'+inf'}}}]}.
+    {test,is_integer,{f,19},[{x,1}]}.
+    {test,is_ge,{f,19},[{tr,{x,1},{t_integer,any}},{integer,0}]}.
+    {test,is_ge,{f,19},[{integer,1023},{tr,{x,1},{t_integer,{0,'+inf'}}}]}.
```

Comment:
I assume that similar test instructions are emitted in OTP 25 as well but with less type information?
I am a bit confused if there are more instructions in OTP 26 than in OTP 25.

Reply:
Yes, it is probably best that I'll show the code for OTP 25 as well.

Reply:
I have pushed a commit that shows more of BEAM instructions for OTP 25.

In order to explain that OTP 25 emits `is_lt` and OTP 26 `is_ge`, I added a new section in this commit: https://github.com/bjorng/erlang-org/pull/2/commits/d4b375671829a41d56f034bbf5e9689c9a6b872a, in which I also show some other optimizations.